ui/pb/progressbar.go

```diff
@@ -0,0 +1,142 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2019 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package pb
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/fatih/color"
+)
+
+const defaultWidth = 40
+const defaultBarColor = color.Faint
+
+// ProgressBar is just a simple thread-safe progressbar implementation with
+// callbacks.
+type ProgressBar struct {
+	mutex sync.RWMutex
+	width int
+	color *color.Color
+
+	left     func() string
+	progress func() (progress float64, right string)
+	hijack   func() string
+}
+
+// ProgressBarOption is used for helper functions that modify the progressbar
+// parameters, either in the constructor or via the Modify() method.
+type ProgressBarOption func(*ProgressBar)
+
+// WithLeft modifies the function that returns the left progressbar padding.
+func WithLeft(left func() string) ProgressBarOption {
+	return func(pb *ProgressBar) { pb.left = left }
+}
+
+// WithConstLeft sets the left progressbar padding to the supplied const.
+func WithConstLeft(left string) ProgressBarOption {
+	return func(pb *ProgressBar) {
+		pb.left = func() string { return left }
+	}
+}
+
+// WithProgress modifies the progress calculation function.
+func WithProgress(progress func() (float64, string)) ProgressBarOption {
+	return func(pb *ProgressBar) { pb.progress = progress }
+}
+
+// WithConstProgress sets the progress and right padding to the supplied consts.
+func WithConstProgress(progress float64, right string) ProgressBarOption {
+	return func(pb *ProgressBar) {
+		pb.progress = func() (float64, string) { return progress, right }
+	}
+}
+
+// WithHijack replaces the progressbar String function with the argument.
+func WithHijack(hijack func() string) ProgressBarOption {
+	return func(pb *ProgressBar) { pb.hijack = hijack }
+}
+
+// New creates and initializes a new ProgressBar struct, calling all of the
+// supplied options
+func New(options ...ProgressBarOption) *ProgressBar {
+	pb := &ProgressBar{
+		mutex: sync.RWMutex{},
+		width: defaultWidth,
+		color: color.New(defaultBarColor),
+	}
+	pb.Modify(options...)
+	return pb
+}
+
+// Modify changes the progressbar options in a thread-safe way.
+func (pb *ProgressBar) Modify(options ...ProgressBarOption) {
```

Comment:
I find the usage of both this and FunctionOptions confusing. I don't see any place where we have multiple options set at the same time. IMHO we can just make the options lock the progressbar themselves

Reply:
> I find the usage of both this and FunctionOptions confusing.

Do you mean `ProgressBarOption` by `FunctionOptions`? And in what way is it confusing? I got the idea for this type of option setting from the [`mpb` lib](https://godoc.org/github.com/vbauerster/mpb#BarOption) and I rather like how flexible and at the same time clear it is what's happening. 

>  I don't see any place where we have multiple options set at the same time.
We don't right now, but I have yet to implement the progressbar alignment and sizing improvements, so we might. And in any case, it won't actually make the usage any simpler, it will basically save 2 lines of code...

> IMHO we can just make the options lock the progressbar themselves
Why? Instead of having the locking in the central place that we need it in only when we need it, we'd spread it across several different functions.

Reply:
@cuonglm, @imiric, thoughts and opinions regarding :arrow_up: ?

Reply:
Resolving this due to inactivity